akka-http/src/main/scala/akka/http/server/directives/CodingDirectives.scala

```diff
@@ -16,46 +18,64 @@ trait CodingDirectives {
   import MiscDirectives._
   import RouteDirectives._
   import CodingDirectives._
+  import HeaderDirectives._
 
   // encoding
 
   /**
-   * Wraps its inner Route with encoding support using the given Encoder.
-   */
-  def encodeResponse(encoder: Encoder): Directive0 =
-    responseEncodingAccepted(encoder.encoding) &
-      mapResponse(encoder.encode(_)) &
-      cancelRejections(classOf[UnacceptedResponseEncodingRejection])
-
-  /**
    * Rejects the request with an UnacceptedResponseEncodingRejection
    * if the given encoding is not accepted for the response.
    */
   def responseEncodingAccepted(encoding: HttpEncoding): Directive0 =
     extract(_.request.isEncodingAccepted(encoding))
-      .flatMap(if (_) pass else reject(UnacceptedResponseEncodingRejection(encoding)))
+      .flatMap(if (_) pass else reject(UnacceptedResponseEncodingRejection(Set(encoding))))
 
   /**
-   * Wraps its inner Route with response compression, using the specified
-   * encoders in the given order of preference.
-   * If no encoders are specifically given Gzip, Deflate and NoEncoding
-   * are used in this order, depending on what the client accepts.
+   * Wraps its inner Route with response encoding, using one of the specified
+   * encoders negotiated with the encodings the client accepts.
+   *
+   * The negotiation logic is defined by http://tools.ietf.org/html/rfc7231#section-5.3.4.
    */
-  def compressResponse(encoders: Encoder*): Directive0 =
-    theseOrDefault(encoders).map(encodeResponse(_)).reduce(_ | _)
+
+  def encodeResponseWith(encoders: Encoder*): Directive0 = _encodeResponse(immutable.Seq(encoders: _*))
 
   /**
-   * Wraps its inner Route with response compression if and only if the client
-   * specifically requests compression with an `Accept-Encoding` header.
+   * Wraps its inner Route with response compression unless the client denied compression with an
+   * `Accept-Encoding` header.
    */
-  def compressResponseIfRequested(): Directive0 = compressResponse(NoCoding, Gzip, Deflate)
+  def encodeResponse: Directive0 = _encodeResponse(DefaultCoders)
```

Comment:
Ok, gentlemen, my preference for "compress-by-default" as not strong enough to warrant a long argument with you, so let's make the `encodeReponse` directive exhibit "compress-if-no-other-preference" behavior only when an `Accept-Encoding` request header is present.
In this case I don't think we really need another high-level directive (like `compressResponse` or similar), a low-level `encodeResponseWith(first: Encoder, more: Encoder*)` should be all that's required.

To recap:

```scala
// encode response with `gzip` or `deflate` if an `Accept-Encoding` header is present and
// the client doesn't prohibit compression (`q=0`) or expresses higher preference for `identity`;
// if no `Accept-Encoding` header is present default to `identity`
def encodeReponse: Directive0 = ...

// encode response with the one of the given encoders,
// which the client expresses highest preference for;
// if client preference is equal (e.g. because there is no `Accept-Encoding` header)
// the Encoder listed first takes precedence;
// if no Encoder is acceptable, respond with `406 Not Acceptable`
def encodeResponseWith(first: Encoder, more: Encoder*): Directive0 = ...
```

WDYT?